### ISSUE TITLE
typing support for public capabilities of Greenwood

### ISSUE THREAD

**Issue by thescientist13** [ProjectEvergreen/greenwood#1250]
## Summary

While Greenwood is a build tool without much of a public API other than its CLI and plugins, there are some areas that should probably provide typing to users through JSDocs, maybe TS but I'm not super keen on TypeScript in libraries.

## Details
Most obvious ones that come to mind are:

1. [_greenwood.config.js_](https://greenwoodjs.dev/docs/reference/configuration/), even if just [JS Docs style](https://nextjs.org/docs/pages/building-your-application/configuring/typescript#type-checking-nextconfigjs), ex.
    ```js
    /** @type {import('@greenwood/cli').GreenwoodConfig} */
    export default {
      /* config options here */
    }
    ```
1. [Plugins API](https://greenwoodjs.dev/docs/reference/plugins-api/) (and get rid of that silly `ResourceInterface` lol)
    - https://github.com/ProjectEvergreen/greenwood/issues/1296
1. [Content as Data](https://greenwoodjs.dev/docs/content-as-data/) - like the data client
1. others?

We would also want to make sure we validate these through CI somehow, ideally using these config files in our own projects and / or test cases.

----

Raised in https://github.com/ProjectEvergreen/greenwood/discussions/1292#discussioncomment-11147829, a good point was raised about _authoring_ Greenwood plugins (and probably also _greenwood.config.js_ in TypeScript.  Might need to make another issue for that once we can land this one first, unless its easy to do it all in one go with something like [**sucrase**](https://www.npmjs.com/package/sucrase), which we are using now in WCC.

**Comment by lschierer** [comment]
I do like writing my config files in typescript, with typing support, because it helps avoid problems where you write an invalid configuration file and unexpected things do and do not work.  I'm not sure which is worse, when a program fails to start on an invalid config file and you have no clue why or where in the file you go something wrong because the file is entirely untyped, or where it loads a partial file and you can't tell if the config file is wrong or the feature doesn't work as expected.

**Comment by thescientist13** [comment]
Indeed!

This will definitely be looked into as part of the next batch of work we take on (just finishing up a new release line as we speak), and for sure even just typing the configuration file alone would be a really nice quality of life enhancement; be it JS Doc style annotations, or just full on TS support, e.g. _greenwood.config.ts_.  💯
